compaction/time_window_compaction_strategy.cc

```diff
@@ -184,16 +184,23 @@ class classify_by_timestamp {
     };
 };
 
-uint64_t time_window_compaction_strategy::adjust_partition_estimate(const mutation_source_metadata& ms_meta, uint64_t partition_estimate) const {
-    if (!ms_meta.min_timestamp || !ms_meta.max_timestamp) {
-        // Not enough information, we assume the worst
-        return partition_estimate / max_data_segregation_window_count;
+uint64_t time_window_compaction_strategy::adjust_partition_estimate(const mutation_source_metadata& ms_meta, uint64_t partition_estimate, schema_ptr s) const {
+    // If not enough information, we assume the worst
+    auto estimated_window_count = max_data_segregation_window_count;
+    auto default_ttl = std::chrono::duration_cast<std::chrono::seconds>(s->default_time_to_live());
+    bool min_and_max_ts_available = ms_meta.min_timestamp && ms_meta.max_timestamp;
+
+    if (!min_and_max_ts_available && default_ttl.count()) {
+        std::chrono::seconds window_size = _options.get_sstable_window_size();
+
+        estimated_window_count = (default_ttl.count() + window_size.count() - 1) / window_size.count();
+    } else if (min_and_max_ts_available) {
+        auto min_window = get_window_for(_options, *ms_meta.min_timestamp);
+        auto max_window = get_window_for(_options, *ms_meta.max_timestamp);
+        const auto window_size = get_window_size(_options);
+
+        estimated_window_count = (max_window + (window_size - 1) - min_window) / window_size;
     }
-    const auto min_window = get_window_for(_options, *ms_meta.min_timestamp);
-    const auto max_window = get_window_for(_options, *ms_meta.max_timestamp);
-    const auto window_size = get_window_size(_options);
-
-    auto estimated_window_count = (max_window + (window_size - 1) - min_window) / window_size;
 
     return partition_estimate / std::max(1UL, uint64_t(estimated_window_count));
```

Comment:
Let's add protection for division by zero which we may have introduced here.

Reply:
That requires window size to be 0, and it would be a preexisting problem.

Reply:
We're adding protection against bad settings, refusing creation or altering with them.